File: backend-dotnet/Services/InvoiceService.cs

using backend_dotnet.DTO;
using backend_dotnet.Models;
using backend_dotnet.Repositories;
using backend_dotnet.Repository;

namespace backend_dotnet.Services
{
    public class InvoiceService : IInvoiceService
    {
        private readonly IInvoiceRepository invoiceRepository;
        private readonly IInvoiceDetailRepository invoiceDetailRepository;
        private readonly IVehicleModelRepository vehicleModelRepository;
        private readonly IComponentRepository componentRepository;
        private readonly IVehicleDetailRepository vehicleDetailRepository;
        private readonly IUserRepository userRepository;
        private readonly IAlternateComponentRepository alternateComponentRepository;

        public InvoiceService(
            IInvoiceRepository invoiceRepository,
            IInvoiceDetailRepository invoiceDetailRepository,
            IVehicleModelRepository vehicleModelRepository,
            IComponentRepository componentRepository,
            IVehicleDetailRepository vehicleDetailRepository,
            IUserRepository userRepository,
            IAlternateComponentRepository alternateComponentRepository)
        {
            this.invoiceRepository = invoiceRepository;
            this.invoiceDetailRepository = invoiceDetailRepository;
            this.vehicleModelRepository = vehicleModelRepository;
            this.componentRepository = componentRepository;
            this.vehicleDetailRepository = vehicleDetailRepository;
            this.userRepository = userRepository;
            this.alternateComponentRepository = alternateComponentRepository;
        }

        public Invoice GetInvoiceById(int invoiceId)
        {
            Invoice? invoice = invoiceRepository.GetById(invoiceId);

            if (invoice == null)
            {
                throw new Exception("Invoice Not Found");
            }

            return invoice;
        }

        public InvoiceResponseDTO GetInvoice(int invoiceId)
        {
            Invoice? invoice = invoiceRepository.GetById(invoiceId);

            if (invoice == null)
            {
                throw new Exception("Invoice Not Found");
            }

            List<VehicleDetail> vehicleDetails =
                vehicleDetailRepository.GetByModelId(
                    invoice.VehicleModel.ModelId);

            List<InvoiceDetail> details =
                invoiceDetailRepository.GetByInvoiceId(invoiceId);

            List<InvoiceDetailResponseDTO> componentList =
                new List<InvoiceDetailResponseDTO>();

            foreach (VehicleDetail vd in vehicleDetails)
            {
                InvoiceDetailResponseDTO dto =
                    new InvoiceDetailResponseDTO();

                dto.ComponentName =
                    vd.Component.CompName;

                InvoiceDetail? matched = details
                    .FirstOrDefault(d =>
                        d.Component.CompId ==
                        vd.Component.CompId);

                if (matched != null)
                {
                    if (matched.AlternateComponent != null)
                    {
                        dto.SelectedComponent =
                            matched.AlternateComponent.CompName;
                    }
                    else
                    {
                        dto.SelectedComponent = "Default";
                    }

                    dto.DeltaPrice = matched.DeltaPrice;
                }
                else
                {
                    dto.SelectedComponent = "Default";
                    dto.DeltaPrice = 0;
                }

                componentList.Add(dto);
            }

            InvoiceResponseDTO response =
                new InvoiceResponseDTO();

            response.invoiceId = invoice.InvoiceId;
            response.invoiceDate = invoice.InvoiceDate;

            response.username =
                invoice.User.Username;

            response.companyName =
                invoice.User.CompanyName;

            response.companyAddress =
                invoice.User.CompanyAddress;

            response.email =
                invoice.User.Email;

            response.mobile =
                invoice.User.Mobile;

            response.gstNo =
                invoice.User.GstNo;

            response.modelName =
                invoice.VehicleModel.ModelName;

            response.manufacturer =
                invoice.VehicleModel.Manufacturer.ManufacturerName;

            response.segment =
                invoice.VehicleModel.Segment.segmentName;

            response.image =
                invoice.VehicleModel.Image;

            response.quantity =
                invoice.Quantity;

            response.totalAmount =
                invoice.TotalAmount;

            response.tax =
                invoice.Tax;

            response.finalAmount =
                invoice.FinalAmount;

            response.components =
                componentList;

            return response;
        }

        public Invoice SaveInvoice(InvoiceRequestDTO request)
        {
            VehicleModel? model =
                vehicleModelRepository.GetById(request.modelId);

            if (model == null)
            {
                throw new Exception("Vehicle Model Not Found");
            }

            User? user =
                userRepository.GetById(request.userId);

            if (user == null)
            {
                throw new Exception("User Not Found");
            }

            Invoice invoice =
                new Invoice();

            invoice.InvoiceDate =
                DateTime.Now;

            invoice.User = user;
            invoice.VehicleModel = model;

            invoice.Quantity =
                request.quantity;

            invoice.TotalAmount =
                request.totalAmount;

            invoice.Tax =
                request.tax;

            invoice.FinalAmount =
                request.finalAmount;

            Invoice savedInvoice =
                invoiceRepository.Save(invoice);

            foreach (SelectedComponentDTO detailRequest
                     in request.selectedComponents)
            {
                InvoiceDetail detail =
                    new InvoiceDetail();

                detail.Invoice =
                    savedInvoice;

                Component? component =
                    componentRepository.GetById(
                        detailRequest.componentId);

                if (component == null)
                {
                    throw new Exception("Component Not Found");
                }

                detail.Component =
                    component;

                if (detailRequest.alternateComponentId != 0)
                {
                    Component? alternate =
                        componentRepository.GetById(
                            detailRequest.alternateComponentId);

                    if (alternate == null)
                    {
                        throw new Exception(
                            "Alternate Component Not Found");
                    }

                    detail.AlternateComponent =
                        alternate;

                    AlternateComponent? alt =
                        alternateComponentRepository
                            .FindByAlternateComponent_CompIdAndComponent_CompIdAndModel_ModelId(
                                detailRequest.alternateComponentId,
                                detailRequest.componentId,
                                request.modelId);

                    if (alt == null)
                    {
                        throw new Exception(
                            "Alternate Mapping Not Found");
                    }

                    detail.DeltaPrice =
                        alt.DeltaPrice;
                }
                else
                {
                    detail.DeltaPrice = 0;
                }

                invoiceDetailRepository.Save(detail);
            }

            return savedInvoice;
        }
    }
}

File: backend/java/Vehicle_Configurator/src/main/java/com/example/demo/services/InvoiceService.java
package com.example.demo.services;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import com.example.demo.dto.InvoiceDetailResponseDTO;
import com.example.demo.dto.InvoiceRequestDTO;
import com.example.demo.dto.InvoiceResponseDTO;
import com.example.demo.dto.SelectedComponentDTO;
import com.example.demo.entities.AlternateComponent;
import com.example.demo.entities.Component;
import com.example.demo.entities.Invoice;
import com.example.demo.entities.InvoiceDetail;
import com.example.demo.entities.User;
import com.example.demo.entities.VehicleDetail;
import com.example.demo.entities.VehicleModel;
import com.example.demo.repositories.AlternateComponentRepository;
import com.example.demo.repositories.ComponentRepository;
import com.example.demo.repositories.InvoiceDetailRepository;
import com.example.demo.repositories.InvoiceRepository;
import com.example.demo.repositories.UserRepository;
import com.example.demo.repositories.VehicleDetailRepository;
import com.example.demo.repositories.VehicleModelRepository;

@Service
public class InvoiceService {

    @Autowired
    private InvoiceRepository invoiceRepository;

    @Autowired
    private InvoiceDetailRepository invoiceDetailRepository;

    @Autowired
    private VehicleModelRepository vehicleModelRepository;

    @Autowired
    private ComponentRepository componentRepository;
    @Autowired
    private VehicleDetailRepository vehicleDetailRepository;
    @Autowired
    private UserRepository userRepository;
    @Autowired
    private AlternateComponentRepository alternateComponentRepository;
    public Invoice getInvoiceById(int invoiceId) {

        return invoiceRepository
                .findById(invoiceId)
                .orElseThrow(() ->
                        new RuntimeException("Invoice Not Found"));

    }
    public InvoiceResponseDTO getInvoice(int invoiceId) {

        Invoice invoice = invoiceRepository.findById(invoiceId)
                .orElseThrow(() -> new RuntimeException("Invoice Not Found"));
        List<VehicleDetail> vehicleDetails =
                vehicleDetailRepository.findByModel_ModelId(
                        invoice.getVehicleModel().getModelId());

        List<InvoiceDetail> details =
                invoiceDetailRepository.findByInvoiceInvoiceId(invoiceId);
        List<InvoiceDetailResponseDTO> componentList =
                new ArrayList<>();

        for (VehicleDetail vd : vehicleDetails) {

            InvoiceDetailResponseDTO dto = new InvoiceDetailResponseDTO();

            dto.setComponentName(vd.getComponent().getCompName());

            InvoiceDetail matched = details.stream()
                    .filter(d -> d.getComponent().getCompId()
                            == vd.getComponent().getCompId())
                    .findFirst()
                    .orElse(null);

            if (matched != null) {

                if (matched.getAlternateComponent() != null) {

                    dto.setSelectedComponent(
                            matched.getAlternateComponent().getCompName());

                } else {

                    dto.setSelectedComponent("Default");

                }

                dto.setDeltaPrice(matched.getDeltaPrice());

            } else {

                dto.setSelectedComponent("Default");
                dto.setDeltaPrice(0);

            }

            componentList.add(dto);
        }
        InvoiceResponseDTO response =
                new InvoiceResponseDTO();

        response.setInvoiceId(invoice.getInvoiceId());

        response.setInvoiceDate(invoice.getInvoiceDate());

        response.setUsername(
                invoice.getUser().getUsername());

        response.setCompanyName(
                invoice.getUser().getCompanyName());

        response.setCompanyAddress(
                invoice.getUser().getCompanyAddress());

        response.setEmail(
                invoice.getUser().getEmail());

        response.setMobile(
                invoice.getUser().getMobile());

        response.setGstNo(
                invoice.getUser().getGstNo());

        response.setModelName(
                invoice.getVehicleModel().getModelName());

        response.setManufacturer(
                invoice.getVehicleModel()
                .getManufacturer()
                .getManufacturerName());

        response.setSegment(
                invoice.getVehicleModel()
                .getSegment()
                .getSegmentName());

        response.setImage(
                invoice.getVehicleModel().getImage());

        response.setQuantity(invoice.getQuantity());

        response.setTotalAmount(invoice.getTotalAmount());

        response.setTax(invoice.getTax());

        response.setFinalAmount(invoice.getFinalAmount());

        response.setComponents(componentList);

        return response;

    }
    public Invoice saveInvoice(InvoiceRequestDTO request) {


        // Fetch Vehicle Model
        VehicleModel model = vehicleModelRepository
                .findById(request.getModelId())
                .orElseThrow(() -> new RuntimeException("Vehicle Model Not Found"));


        // Fetch User
        User user = userRepository
                .findById(request.getUserId())
                .orElseThrow(() -> new RuntimeException("User Not Found"));


        // Create Invoice
        Invoice invoice = new Invoice();

        invoice.setInvoiceDate(LocalDateTime.now());
        invoice.setUser(user);
        invoice.setVehicleModel(model);

        invoice.setQuantity(request.getQuantity());
        invoice.setTotalAmount(request.getTotalAmount());
        invoice.setTax(request.getTax());
        invoice.setFinalAmount(request.getFinalAmount());


        // Save Invoice First
        Invoice savedInvoice = invoiceRepository.save(invoice);



        // Save Invoice Details

        for (SelectedComponentDTO detailRequest : request.getSelectedComponents()) {


            InvoiceDetail detail = new InvoiceDetail();

            detail.setInvoice(savedInvoice);



            Component component = componentRepository
                    .findById(detailRequest.getComponentId())
                    .orElseThrow(() -> new RuntimeException("Component Not Found"));


            detail.setComponent(component);



            if(detailRequest.getAlternateComponentId() != 0) {


                Component alternate = componentRepository
                        .findById(detailRequest.getAlternateComponentId())
                        .orElseThrow(() -> 
                            new RuntimeException("Alternate Component Not Found"));


                detail.setAlternateComponent(alternate);

                AlternateComponent alt =
                		alternateComponentRepository
                		.findByAlternateComponent_CompIdAndComponent_CompIdAndModel_ModelId(
                		        detailRequest.getAlternateComponentId(),
                		        detailRequest.getComponentId(),
                		        request.getModelId()
                		);

                		if (alt == null) {
                		    throw new RuntimeException("Alternate Mapping Not Found");
                		}

                		detail.setDeltaPrice(alt.getDeltaPrice());
            }
            else {

                detail.setDeltaPrice(0);

            }
            invoiceDetailRepository.save(detail);

        }


        return savedInvoice;
    }

}